src/porepy/numerics/contact_mechanics/contact_conditions.py

```diff
@@ -0,0 +1,463 @@
+        delta = min(self._l2(-Tt) / bf, 1)
+
+        if alpha < 0:
+            beta = 1 / (1 - alpha * delta)
+        else:
+            beta = 1
+
+        # The expression (I - beta * M)^-1
+        IdM_inv = np.linalg.inv(Id - beta * coeff_M)
+
+        v = IdM_inv.dot(-Tt + cut) / self._l2(-Tt + cut)
+
+        return c * (IdM_inv - Id), -IdM_inv.dot(self._hf(Tt, cut, bf)), v
+
+    def _l2(self, x):
+        x = np.atleast_2d(x)
+        return np.sqrt(np.sum(x ** 2, axis=0))
+
+
+def set_projections(gb):
+    """ Define a local coordinate system, and projection matrices, for all
+    grids of co-dimension 1.
+
+    The function adds one item to the data dictionary of all GridBucket edges
+    that neighbors a co-dimension 1 grid, defined as:
+        key: tangential_normal_projection, value: pp.TangentialNormalProjection
+            provides projection to the surface of the lower-dimensional grid
+
+    Note that grids of co-dimension 2 and higher are ignored in this construction,
+    as we do not plan to do contact mechanics on these objects.
+
+    It is assumed that the surface is planar.
+
+    """
+
+    ambient_dim = gb.dim_max()
+
+    # Information on the vector normal to the surface is not available directly
+    # from the surface grid (it could be constructed from the surface geometry,
+    # which spans the tangential plane). We instead get the normal vector from
+    # the adjacent higher dimensional grid.
+    # We therefore access the grids via the edges of the mixed-dimensional grid.
+    for e, d_m in gb.edges():
+
+        mg = d_m["mortar_grid"]
+        # Only consider edges where the lower-dimensional neighbor is of co-dimension 1
+        if not mg.dim == ambient_dim - 1:
+            continue
+
+        # Neigboring grids
+        _, g_h = gb.nodes_of_edge(e)
+
+        # Find faces of the higher dimensional grid that coincide with the mortar
+        # grid. Go via the master to mortar projection
+        # Convert matrix to csr, then the relevant face indices are found from
+        # the row indices
+        faces_on_surface = mg.master_to_mortar_int().tocsr().indices
+
+        # Find out whether the boundary faces have outwards pointing normal vectors
+        # Negative sign implies that the normal vector points inwards.
+        sgn = g_h.sign_of_faces(faces_on_surface)
+
+        # Unit normal vector
+        unit_normal = g_h.face_normals[: g_h.dim] / g_h.face_areas
+        # Ensure all normal vectors on the relevant surface points outwards
+        unit_normal[:, faces_on_surface] *= sgn
+
+        # Now we need to pick out a normal vector of the higher dimensional grid
+        # which coincides with this mortar grid. This could probably have been
+        # done with face tags, but we instead project the normal vectors onto the
+        # mortar grid to kill off all irrelevant grids.
+        outwards_unit_vector_mortar = mg.master_to_mortar_int().dot(unit_normal.T).T
```

Comment:
Actually no - we use a single normal vector for the entire fracture (some lines below). It will crash for non-planar fractures, but that will have to be another day.

You are right the code and documentation was confusing, I will update.